src/v/kafka/server/group_metadata.cc

```diff
@@ -390,11 +392,61 @@ group_metadata_serializer make_backward_compatible_serializer() {
     };
     return group_metadata_serializer(std::make_unique<impl>());
 }
+iobuf maybe_unwrap_from_iobuf(iobuf buffer) {
+    /*
+     * Previously in redpanda we had an error leading to wrapping
+     * tombstone record keys with an iobuf serialization envelope. In order to
+     * provide compatibility and be able to read keys wrapped with an additional
+     * iobuf we introduce a check leveraging serialization format.
+     *
+     * In Kafka the only allowed values for group metadata versions are 0, 1,
+     * and 2. Group metadata use big endian encoding hence the first two bytes
+     * of a serialized group metadata key contains information about a version.
+     * Both offset_metadata_key and group_metadata_key first serialized field
+     * is a `group_id` string which is encoded as 2 bytes of size and then
+     * array of characters. This way 4 first bytes of encoded key are:
+     *
+     * b0 - denotes LSB
+     * bN - denotes N byte indexed from LSB to MSB (LSB has index 0)
+     *
+     * Key bytes (big endian):
+     *
+     *      |version_b1|version_b0|size_b1|size_b0|...
+     *
+     * When key struct is wrapped with an iobuf serialization it is prefixed
+     * with iobuf size which is serialized as 4 bytes of a size (in a little
+     * endian encoding) and then the iobuf content
+     *
+     * iobuf size bytes (little endian):
+     *
+     *      |size_b0|size_b1|size_b2|size_b3|...
+     *
+     * To check if metadata key was wrapped with an iobuf we peek first 4 bytes
+     * to try to decode iobuf size store in little endian format if returned
+     * size is equal to size of remaining key buffer size it means that we need
+     * to unwrap key from an iobuf.
+     *
+     *
+     * TODO: remove this code in future releases
+     */
+
+    iobuf_parser parser(buffer.copy());
+    // peek first 4-byte
+    auto deserialized_size = reflection::from_iobuf<int32_t>(parser.peek(4));
+
+    if (unlikely(
+          buffer.size_bytes() == static_cast<size_t>(deserialized_size) + 4)) {
```

Comment:
It seems pretty unlikely that this condition could give us a false-positive, i.e. that `size_b0 == version_b1 && size_b1 == verstion_b0`, etc. for the size of the buffer. Do we have a simple proof / argument though?

Reply:
Added a proof that the approach is valid in cover letter.

Reply:
Wow.. Great response. Nice work.